distributed/tests/test_executor.py

```diff
@@ -3202,6 +3204,94 @@ def f():
                 future = e.submit(f)
                 assert future.result() == 1
 
+
+@gen_cluster(executor=False)
+def test_publish_simple(s, a, b):
+    e = Executor((s.ip, s.port), start=False)
+    yield e._start()
+    f = Executor((s.ip, s.port), start=False)
+    yield f._start()
+
+    data = yield e._scatter(range(3))
+    yield e._publish_dataset(data, 'data')
+    assert 'data' in s.published_data
+    assert e.id in s.published_data['data']['clients']
+
+    result = yield e.scheduler.list_datasets()
+    assert result == ['data']
+
+    result = yield f.scheduler.list_datasets()
+    assert result == ['data']
+
+
+@gen_cluster(executor=False)
+def test_publish_roundtrip(s, a, b):
+    e = Executor((s.ip, s.port), start=False)
+    yield e._start()
+    f = Executor((s.ip, s.port), start=False)
+    yield f._start()
+
+    data = yield e._scatter([0, 1, 2])
+    yield e._publish_dataset(data, 'data')
+
+    assert 'published-data' in s.who_wants[data[0].key]
+    result = yield f._get_dataset(name='data')
+
+    assert len(result) == len(data)
+    out = yield f._gather(result)
+    assert out == [0, 1, 2]
+
+    yield e._shutdown()
+    assert e.id not in s.published_data['data']['clients']
+
+    yield f._shutdown()
+    assert 'data' not in s.published_data
+
+    assert 'published-data' not in s.who_wants[data[0].key]
+
+
+@gen_cluster(executor=False)
+def test_unpublish(s, a, b):
+    e = Executor((s.ip, s.port), start=False)
+    yield e._start()
+    data = yield e._scatter([0, 1, 2])
+    yield e._publish_dataset(data, 'data')
+
+    key = data[0].key
+    del data
+
+    yield e.scheduler.unpublish_dataset(name='data')
+
+    assert 'data' not in s.published_data
+    assert not s.who_wants[key]
```

Comment:
Doesn't the yield on unpublish ensure that that method runs in its entirety before returning control? It directly calls client_releases_keys, which edits who_wants - not by state transition or secondary action.